### PR TITLE
Support passing parameters to dataset definition modules

### PR DESCRIPTION
**Pull request by inglesp**

### CODE REVIEW

**Review by evansd — COMMENTED** [review]
I'm not convinced by this approach for a couple of reasons. The main one is that using `parse_known_kwargs()` means that incorrect arguments to databuilder now pass silently by default. If the user happens to implement strict parsing of the remaining args in their dataset definition then they'll get picked up there. But we can't assume that will happen: I imagine most users will do no parsing at all.

The other is that importing a shadow copy of `argparse` feels a bit heavyweight and leaves it obscure what things are just standard Python and what's Data Builder magic.

I wonder if instead we could use the convention that `--` marks the end of argument parsing and insist that user arguments go after that. So `databuilder.__main__` could split the args on the first `--` and use strict parsing on everything up to that. And then user arguments could be parsed using something like:
```py
from argparse import ArgumentParser

from databuilder import user_args


parser = ArgumentParser()
parser.add_argument("--year", type=int)
args = parser.parse_args(user_args)
```

File: databuilder/__main__.py
Comment (evansd) [inline databuilder/__main__.py:65]: If `--` is not in args then `user_args` will not be changed from `None` which means that `parser.parse_args(user_args)` will end up parsing the whole of `sys.argv[1:]` which I don't think is what we want.

File: databuilder/__init__.py
Comment (evansd) [inline databuilder/__init__.py:7]: I think it would be useful to maintain the property that dataset definitions are executable without error directly in Python i.e. by running
```sh
python analyis/dataset_definition.py
```

As that doesn't go through `databuilder.__main__` it won't end up setting `user_args`.

I think it would be reasonable to populate `user_args` at import time, given that `sys.argv` is. But maybe it would be cleaner to make it a function like `get_user_args()` e.g.
```py
def split_cmd_and_user_args(args):
    if args is None:
        args = sys.argv[1:]
    if "--" in args:
        separator_index = args.index("--")
        cmd_args = args[:separator_index]
        user_args = args[separator_index + 1:]
    else:
        cmd_args = args
        user_args = []
    return cmd_args, user_args


def get_cmd_args(args=None):
    return split_cmd_and_user_args(args)[0]


def get_user_args(args=None):
    return split_cmd_and_user_args(args)[1]
```

Comment (inglesp) [reply databuilder/__init__.py:7]: Thanks.  Would setting `user_args = []` in databuilder/__init__.py do the trick?

Comment (evansd) [reply databuilder/__init__.py:7]: It would mean you had no way of passing arguments to the dataset definition other than by invoking it via databuilder. I think it would be nice if you could do:
```
python analysis/dataset_definition.py -- foo --bar=baz
```
And have it all work correctly.

I mean, that's obviously not essential. But given it's not hard to achieve either it seems worthwhile.

File: databuilder/main.py
Comment (inglesp) [inline databuilder/main.py:243]: The extra plumbing is worth it to get hold of `str(module_path)` here.

**Review by inglesp — COMMENTED** [review]
I can't approve this, presumably because I opened it, but I would approve if I could!

**Review by evansd — APPROVED** [review]
Thanks! I'll approve my own code on your behalf then.

**Pull request merged**
